keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,346 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The service controller in the Kubernetes cloud controller manager (KCCM)
+configures the load balancer and its corresponding health check (HC) following
+service related events. The configured health check is then used by the load
+balancer as to determine which instances are candidates for traffic load
+balancing. For certain cloud providers (GCP): the KCCM configures the HC to
+target the Kubernetes service proxy for this information. In this KEP we will
+focus on Kube-proxy, since it's the only service proxy under the responsibility
+of the Kubernetes project.
+
+We can define two classes of services for what concerns the HC:
+
+- A) services which have `externalTrafficPolicy: Cluster`
+- B) services which have `externalTrafficPolicy: Local`
+
+For class A) services: Kube-proxy currently returns an answer following its
+`healthz` state. For class B) services: Kube-proxy only reports if the service
+for which the load balancer was created for, has a `Ready` endpoint running on
+the node.
+
+This KEP proposes two changes:
+
+1. That Kube-proxy's includes the `healthz` state AND its current answer w.r.t
+  the local endpoints, when it answers to the HC for class B) services.
+
+2. That Kube-proxy enables connection draining for terminating Nodes. This is to
+  be done by Kube-proxy by inspecting a field on the Node object which indicates
+  that the "node is terminating/deleting" and when seen, starts failing the HC.
+
+```
+<<[UNRESOLVED HC-linked-service-fields ]>>
+We need to agree on this field. `deletionTimestamp` is the perfect indicator for
+this, but....the Node object does not have a finalizer, so that wouldn't help.
+@thockin advocates for `.spec.unschedulable`, but I am inclined to say "that
+doesn't indicate if a Node is going away". `unschedulable` just says that no
+workloads can be assigned to the Node anymore. It is true that it's usually
+followed up by a drain...but I am not sure it's good to implement something
+which can easily kill ingress connectivity to a Node / cluster by simply doing
+`kubectl cordon node ...` (It would kill ingress connectivity to the entire
+cluster if the ingress controller is running on the node which is being cordoned)
+
+Maybe we need a KEP for a finalizer on the Node object (*troll face)
+<<[/UNRESOLVED]>>
+```
+
+```
+<<[UNRESOLVED HC-linked-service-fields ]>>
+Figure out which other cloud-providers configure the HC to target the service
```

Comment:
we should talk-up this change to make sure anyone who needs to use this knows it is here.  I don't think this is UNRESOLVED though :)

Reply:
OK, I misunderstood what "UNRESOLVED" indicates on KEPs. I thought it's "unresolved for now, but needs to get resolved before the KEP merges, but I have no clue so maybe someone else can chime in"

Reply:
That is what UNRESOLVED means.  Maybe I misunderstand is this point "we need to tell people this is coming and encourage them to use it" or "we can't proceed witht he KEP until we audit the top N other providers"?

I don't think we need to block the KEP - we can merge this and also reach out in parallel.

Reply:
> "we can't proceed witht he KEP until we audit the top N other providers"?

I meant that. Like a TODO for myself to verify that...but for out-of-tree providers I might not even be able to look (if it's closed source). 

> I don't think we need to block the KEP - we can merge this and also reach out in parallel.

OK, let me mention that on the KEP

Reply:
> out-of-tree providers I

it's open source

https://github.com/kubernetes/cloud-provider-gcp
https://github.com/kubernetes/cloud-provider-aws
https://github.com/kubernetes-sigs/cloud-provider-azure

Reply:
I will have a look at those: I quickly just did:  ctrl + f + "10256" on all of those...only GCP turned up. In any case: I mentioning on this KEP that one proposal is also to add official documentation somewhere in Kube which details how all of this works and defines a recommendation for cloud providers to use.

Reply:
ah, one useful thing is this https://grep.app/search?q=10256&filter[path.pattern][0]=k8s&filter[lang][0]=Go

Reply:
and yes, it seems this is a GCP only thing :)